### ISSUE TITLE
DictList.insertBefore only populates the Dict and not the list

### ISSUE THREAD

**Issue by rosensama** [Gizra/elm-dictlist#16]
If I use DictList.insertBefore to maintain the container contents, the dictionary portion is populated as expected, but the list portion remains empty.   I believe a symptom of this is that DictList.values returns an empty list regardless of DictList Dict population.

version 2.1.0

**Comment by rgrempel** [comment]
Sorry not to respond sooner!

I'll take a look and see what might be causing this.

If you still have that code around, what specific kind of `DictList` were you using? E.g. was it a `DictList String String` ... what were the key and value types? And was it literally a `DictList`, or an `EveryDictList` or `AllDictList`? (At first glance, I can see a possible problem that would affect some kinds of `AllDictList` but probably not the others).

**Comment by rgrempel** [comment]
Sorry not to respond sooner!

I'll take a look and see what might be causing this.

If you still have that code around, what specific kind of `DictList` were you using? E.g. was it a `DictList String String` ... what were the key and value types? And was it literally a `DictList`, or an `EveryDictList` or `AllDictList`? (At first glance, I can see a possible problem that would affect some kinds of `AllDictList` but probably not the others).

**Comment by rgrempel** [comment]
Sorry not to respond sooner!

I'll take a look and see what might be causing this.

If you still have that code around, what specific kind of `DictList` were you using? E.g. was it a `DictList String String` ... what were the key and value types? And was it literally a `DictList`, or an `EveryDictList` or `AllDictList`? (At first glance, I can see a possible problem that would affect some kinds of `AllDictList` but probably not the others).

**Comment by rgrempel** [comment]
Sorry not to respond sooner!

I'll take a look and see what might be causing this.

If you still have that code around, what specific kind of `DictList` were you using? E.g. was it a `DictList String String` ... what were the key and value types? And was it literally a `DictList`, or an `EveryDictList` or `AllDictList`? (At first glance, I can see a possible problem that would affect some kinds of `AllDictList` but probably not the others).

**Comment by rosensama** [comment]
Thanks for taking a look.  

I'm using something like this.  I wound up switching to `DictList.cons` instead of `DictList.insertBefore`.  Sorry I didn't send a simple repro at the time.  

```elm
type alias Things = 
    DictList Int Thing

type alias Thing = 
    { a : Int,
    , b : Int,
    , c : Float,
    , d : Int,
    , e : Date
    }

DictList.insertBefore x y thing things

-- this would return nothing
DictList.values things
```

**Comment by rgrempel** [comment]
My apologies for the long delay in following up on this!

I tried to write a few tests that might reproduce the problem, but I can't seem to get them to fail. Here's what I tried:

https://github.com/Gizra/elm-dictlist/pull/19/commits/c683147bff1682ab82cad67dd1bd99dba93eeaad#diff-41fce83a716aea95f58634b46dbf2defR877

**Comment by rgrempel** [comment]
I do see a possible problem with an `AllDictList` that is using a non-comparable value as the key (with a custom comparator function) ... but in your case the key was an `Int`?

In any event, I'll try to reproduce the problem with `AllDictList` and non-comparable keys.

**Comment by rgrempel** [comment]
Yes, I have been able to reproduce a problem on #19 in the following circumstances:

- You are using `AllDict`
- Your custom comparator function isn't consistent with what `==` returns on your key type ... that is, it returns `False` when your comparator function returns `EQ`, or vice versa.

That kind of comparator function is quite legitimate, since it's what you'd need for key types that have distinct internal representations that ought to be considered equal. So, basically I can't trust `==` for keys in the implementation of `AllDictList`, and there are a variety of places where I've being doing that when maintaining the list.

I'll work on a fix for this. It is related to your original problem ... but I don't think it can literally be what your original problem was. You were using a `DictList Int Thing`, and an `Int` key would be consistent as between `==` and the comparator function. So, I'm still not sure what the source of your original problem actually was.

**Comment by rrosenfeld** [comment]
I just tried to repro now in 2.1.1 and cannot do so either.   Thanks for looking.   If I come across it again I will make sure to collect a repro.

**Comment by rgrempel** [comment]
Thanks -- your report did clue me in to another problem!

**Comment by rgrempel** [comment]
I think I have found a scenario in which your original problem with `insertBefore` could occur ... you would have to call `insertBefore` where the new key was equal to the key you were inserting before, and the key wasn't currently being used. So, in the example you gave:

```elm
DictList.insertBefore x y thing things
```

... there was a problem if `x == y` and `DictList.member x things == False`. In that case, you would get just the behaviour you observed ... added to the dict, but not the list.

**Comment by rgrempel** [comment]
Published in version 2.1.2.
